smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,475 @@
+
+        # successive halving parameters
+        self._init_sh_params(min_budget, max_budget, eta, init_chal)
+
+        # adaptive capping
+        if not self.cutoff_as_budget and self.instance_order != 'shuffle' and self.run_obj_time:
+            self.adaptive_capping = True
+        else:
+            self.adaptive_capping = False
+
+    def _init_sh_params(self, min_budget: float,
+                        max_budget: float,
+                        eta: float,
+                        init_chal: int):
+        """
+        initialize Successive Halving parameters
+
+        Parameters
+        ----------
+        min_budget : float
+            minimum budget allowed for 1 run of successive halving
+        max_budget : float
+            maximum budget allowed for 1 run of successive halving
+        eta : float
+            'halving' factor after each iteration in a successive halving run
+        init_chal : int
+            number of challengers to consider for the initial budget
+        """
+
+        if eta <= 1:
+            raise ValueError('eta must be greater than 1')
+        self.eta = eta
+
+        # BUDGETS
+
+        # - if only 1 instance was provided & quality objective, then use cutoff as budget
+        # - else, use instances as budget
+        if not self.run_obj_time and len(self.instances) <= 1:
+            # budget with cutoff
+            # cannot run successive halving with cutoff as budget if budget limits are not provided!
+            if min_budget is None or \
+                    (max_budget is None and self.cutoff is None):
+                raise ValueError("Successive Halving with runtime-cutoff as budget (i.e., only 1 instance) "
+                                 "requires parameters min_budget and max_budget/cutoff for intensification!")
+
+            # if both cutoff and max_budget are provided, then warn user
+            if self.cutoff is not None and max_budget is not None:
+                warnings.warn('Successive Halving with runtime-cutoff as budget: '
+                              'Both max budget (%d) and runtime-cutoff (%d) were provided. Max budget will be used.' %
+                              (self.max_budget, len(self.instances)))
+
+            self.min_budget = min_budget
+            self.max_budget = max_budget if max_budget else self.cutoff
+            self.cutoff_as_budget = True
+
+        else:
+            # budget with instances
+            if self.run_obj_time and len(self.instances) <= 1:
+                warnings.warn("Successive Halving has objective 'runtime' but only 1 instance-seed pair.")
+            self.min_budget = 1 if min_budget is None else int(min_budget)
+            self.max_budget = len(self.instances) if max_budget is None else int(max_budget)
+            self.cutoff_as_budget = False
+
+            # max budget cannot be greater than number of instance-seed pairs
```

Comment:
The same holds for min budget. Also min budget cannot be larger than max budget. Could you please add tests for that?